### PR TITLE
AIOOBE during search for references

### PR DESCRIPTION
**Pull request by trancexpress**
AIOOBE during search for references

Prevent creating a type argument for a type reference with empty type
arguments. I.e. for a type reference "MyType<>" don't create a type
argument, to prevent an AIOOBE later on during a search.

Fixes: #825

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by trancexpress** [comment]
@stephan-herrmann can you check this change? Is this fix what you meant with your comment? https://bugs.eclipse.org/bugs/show_bug.cgi?id=550067#c22

Or the fix should be elsewhere?

**Comment by gayanper** [comment]
@trancexpress Is this PR complete for merging, If so I can have a look and merge it. if you need time to check with @stephan-herrmann we can wait.

**Comment by trancexpress** [comment]
> @trancexpress Is this PR complete for merging, If so I can have a look and merge it. if you need time to check with @stephan-herrmann we can wait.

Best to wait a few days, I'm not sure if Stephan meant this change exactly with his comment. If there is no answer I'll need to look into the bug myself.

File: org.eclipse.jdt.core/model/org/eclipse/jdt/internal/compiler/parser/SourceTypeConverter.java [hunk 1]
```diff
@@ -569,8 +571,17 @@ private TypeDeclaration convert(SourceType typeHandle, CompilationResult compila
 
 		/* set superclass and superinterfaces */
 		if (typeInfo.getSuperclassName() != null) {
-			type.superclass = createTypeReference(typeInfo.getSuperclassName(), start, end, true /* include generics */);
+			char[] superclassName = typeInfo.getSuperclassName();
+			int n = superclassName.length;
+			boolean emptyTypeArgument = n > 2 && superclassName[n - 2] == '<' && superclassName[n - 1] == '>';
```
Comment (gayanper) [inline org.eclipse.jdt.core/model/org/eclipse/jdt/internal/compiler/parser/SourceTypeConverter.java:576]: Will renaming this as `isDiamond` make more sense ?

**Comment by gayanper** [comment]
Once you have checked what Stephan mentioned, please update the checklist in the PR, then we can merge.

**Comment by trancexpress** [comment]
From my POV this is ready to merge. Please review.

**Review by gayanper — APPROVED** [review]
LGTM

**Comment by gayanper** [comment]
Thanks @trancexpress

**Pull request merged**